code/modules/research/rdconsole.dm

```diff
@@ -250,60 +250,63 @@ proc/CallMaterialName(ID)
-							if(linked_lathe) //Also sends salvaged materials to a linked protolathe, if any.
-								for(var/material in linked_destroy.loaded_item.materials)
-									linked_lathe.materials.insert_amount(min((linked_lathe.materials.max_amount - linked_lathe.materials.total_amount), (linked_destroy.loaded_item.materials[material]*(linked_destroy.decon_mod/10))), material)
-								feedback_add_details("item_deconstructed","[linked_destroy.loaded_item.type]")
-							linked_destroy.loaded_item = null
-						else
-							screen = 1.0
-						for(var/obj/I in linked_destroy.contents)
-							for(var/mob/M in I.contents)
-								M.death()
-							if(istype(I,/obj/item/stack/sheet))//Only deconsturcts one sheet at a time instead of the entire stack
-								var/obj/item/stack/sheet/S = I
-								if(S.amount > 1)
-									S.amount--
-									linked_destroy.loaded_item = S
-								else
-									qdel(S)
-									linked_destroy.icon_state = "d_analyzer"
+			linked_destroy.busy = 1
+			screen = 0.1
+			updateUsrDialog()
+			flick("d_analyzer_process", linked_destroy)
+			spawn(24)
+				if(linked_destroy)
+					linked_destroy.busy = 0
+					if(!linked_destroy.loaded_item)
+						usr <<"<span class='danger'>The destructive analyzer appears to be empty.</span>"
+						screen = 1.0
+						return
+
+					for(var/T in temp_tech)
+						var/datum/tech/KT = files.known_tech[T] //For stat logging of high levels
+						if(files.IsTechHigher(T, temp_tech[T]) && KT.level >= 5) //For stat logging of high levels
+							feedback_add_details("high_research_level","[KT][KT.level + 1]") //+1 to show the level which we're about to get
+						files.UpdateTech(T, temp_tech[T])
+
+					if(linked_lathe) //Also sends salvaged materials to a linked protolathe, if any.
+						for(var/material in linked_destroy.loaded_item.materials)
+							linked_lathe.materials.insert_amount(min((linked_lathe.materials.max_amount - linked_lathe.materials.total_amount), (linked_destroy.loaded_item.materials[material]*(linked_destroy.decon_mod/10))), material)
+						feedback_add_details("item_deconstructed","[linked_destroy.loaded_item.type]")
+					linked_destroy.loaded_item = null
+					for(var/obj/I in linked_destroy.contents)
+						for(var/mob/M in I.contents)
+							M.death()
+						if(istype(I,/obj/item/stack/sheet))//Only deconsturcts one sheet at a time instead of the entire stack
+							var/obj/item/stack/sheet/S = I
+							if(S.amount > 1)
+								S.amount--
+								linked_destroy.loaded_item = S
 							else
-								if(!(I in linked_destroy.component_parts))
-									qdel(I)
-									linked_destroy.icon_state = "d_analyzer"
-						use_power(250)
-						updateUsrDialog()
+								qdel(S)
+								linked_destroy.icon_state = "d_analyzer"
+						else
+							if(!(I in linked_destroy.component_parts))
+								qdel(I)
+								linked_destroy.icon_state = "d_analyzer"
+				screen = 1.0
+				use_power(250)
+				updateUsrDialog()

```

Comment:
Code above this is just making the code more simple and preventing some runtimes and lock-up bugs